skimage/morphology/rolling_ball.py

```diff
@@ -0,0 +1,186 @@
+import numpy as np
+from numpy.lib.stride_tricks import as_strided
+
+from skimage.util import invert, view_as_windows
+from ._rolling_ball_cy import apply_kernel, apply_kernel_nan
+
+
+def rolling_ellipsoid(image, kernel_size=(100, 100), intensity_vertex=100,
+                      has_nan=False):
+    """Estimate background intensity using a rolling ellipsoid.
+
+    The rolling ellipsoid algorithm estimates background intensity for a
+    grayscale image in case of uneven exposure. It is a generalization of the
+    frequently used rolling ball algorithm [1]_.
+
+    Parameters
+    ----------
+    image : ndarray
+        The last two dimensions are treated as the columns and rows of the
+        image. The operation is broadcasted along the remaining ones.
```

Comment:
Ah, I think this part is important. So the new behavior allows one to apply the 2D rolling ball algorithm to ND images but the algorithm itself still works with a 2-dimensional ellipsoid right? When reading your comment on "arbitrary ndimage support" I expected ND support that treats all dimensions equally by rolling an ND ellipsoid.

I would have judged the added complexity with using `as_strided` to not be worth it but I'm content to keep this behavior now. Provided you add thorough tests for this. :wink: 

In case you're interested in using a ND-ellipsoid you can look for inspiration to other morphological functions, such as `flood`, `watershed` or `local_maxima` which all use [_offsets_to_raveled_neighbors](https://github.com/scikit-image/scikit-image/blob/1188c33c3defefdc0f672870b5bcc36beaec967f/skimage/morphology/_util.py#L57-L126) under the hood to iterate raveled N-dimensional images. But I'd say that's something out of scope of this current PR. :wink:

Reply:
> Ah, I think this part is important. So the new behavior allows one to apply the 2D rolling ball algorithm to ND images but the algorithm itself still works with a 2-dimensional ellipsoid right?

Yes, this is how I understand the generalization should work.

> I would have judged the added complexity with using as_strided to not be worth it but I'm content to keep this behavior now. Provided you add thorough tests for this. 😉

We can get the relevant windows without copying anything. It generates some unnecessary ones (hence `offsets`), but since the creation is free I don't think we mind that too much. Creating an array of (offset-)integers is still substantially cheaper than copying windows; especially in ND images.

> In case you're interested in using a ND-ellipsoid [...]

I don't understand how that would work, and am very lost the more I think about it. Perhaps you can shed some light on this?

Say we have a stack of images; dunno let's say 60 images from a confocal microscope. Let's say for simplicity that we are talking about a ball with radius `r`. What is the algorithm expected to do? It should grab the volume `B_r([z,y,x])` and raise said ball in the 4th (intensity) dimension until it's surface touches the 3D umbra? 

I can implement that without too much hassle, but does it make sense? Shouldn't background illumination change somewhat significantly as you change the image plane (due to scattering/reflection as light passes through the sample)? Are neighboring stacks related enough to use the minimum across the entire sub-stack?

What about passing in a time-lapse of images? Depending on the experimental setup - time difference + stray light - we can get substantial changes in background illumination and may not want to use information from previous or preceding images.

How would we differentiate between `[plane nrow ncol]` or `[time nrow ncol]` images if a z-stack is indeed a different beast than a sequence of images? Would we want to roll a 4D ellipsoid in a sequence of stacks (`[time pln nrow ncol]`)?

Why are we calling it ND-ellipsoid and not 3D ellipsoid? If above z-stack example is valid, I know how this would generalize to 5D, 6D, or even ND?

Have you encountered such an algorithm before? If so, do you have a name or a link I can look at to understand how it is expected to operate?

Also sorry for the huge wall of questions; it seems like this should be something small, but I can't seem to wrap my head around it, and that somewhat bothers me.

Reply:
For now, I think applying the rolling ball only to two dimensions does make sense. The most common use case for backround subtraction - that I know - is to correct for vignetting (images being brighter in the center than at the edges). Since this is a 2D effect, a 2D correction is totally sufficient.

I believe it would be great if it was possible to specify which dimensions to apply the ball to. For example, the last dimension is often the color dimension and in my case the first dimension is the time dimension.

The more I think about it, the more I get the impression that using a higher dimension ellipsoid could be interesting as well. For example, if you have a time series of moving objects, including the time dimension could help with finding the actual background level and avoid to remove signal. Imagine a bright cell that moves across the spot where you apply the ball. In the 2D case, the ball always "dips" into the center of the cell and removes part of the signal. In the 3D case, the ball will also take into account neighboring frames where the cell has moved away and the ball hits the correct background noise level. 

However, this feature would only be useful, if the radius of the ball could be specified separately for the spatial and for the temporal dimension, because the temporal resolution is not correlated with the spatial resolution.

Also for 3D images, in certain cases it could be an advantage to use a 3D ball. That way, the ball would again find the minimum noise level also in the 3rd dimension. This could be interesting for 3D structures that are large in the x and y dimensions but small in the z dimension.

tl/dr: Ideally, the dimensions that the ellipsoid/spheroid is applied to would be specified by a parameter. If only 2 dimensions are given, the behavior would be as it is now. If more dimensions are given, a higher dimension spheroid would be used.

Reply:
@thawn

> I believe it would be great if it was possible to specify which dimensions to apply the ball to. For example, the last dimension is often the color dimension and in my case the first dimension is the time dimension.

Currently, you can shift the dimensions around with `np.moveaxis` and make it a channel-first image. Then, take the output of `rolling_ellipsoid` and shift the axis back to make it channel-last. 

I agree it would be nice to handle color images without too much fuzz, but as I understand it, the target seems to be gray images only.

>The more I think about it, the more I get the impression that using a higher dimension ellipsoid could be interesting as well. For example, if you have a time series of moving objects, including the time dimension could help with finding the actual background level and avoid to remove signal. Imagine a bright cell that moves across the spot where you apply the ball. In the 2D case, the ball always "dips" into the center of the cell and removes part of the signal. In the 3D case, the ball will also take into account neighboring frames where the cell has moved away and the ball hits the correct background noise level.

If you want to avoid "dips" you can specify a very flat ellipsis and - essentially - perform a rolling_disk. You can't use `intensity_vertex=0` because one can't divide by 0, but `intensity_vertex=1e-6` or so is very possible.

Sidenote: If you have the scenario that the background illumination doesn't change over time, wouldn't it be much simpler to take a picture of the background and then subtract that one from each frame of the video? It will for sure be faster...

> However, this feature would only be useful, if the radius of the ball could be specified separately for the spatial and for the temporal dimension, because the temporal resolution is not correlated with the spatial resolution.

That's what the generalization to an ND ellipsis over an ND ball does :)

> Also for 3D images, in certain cases it could be an advantage to use a 3D ball.

This could be interesting. Do you have a concrete example?

Reply:
>

 >>Also for 3D images, in certain cases it could be an advantage to use a 3D ball.
>This could be interesting. Do you have a concrete example?

Any 3D structures that are large in the x and y dimensions but small in the z dimension.
From cell biology examples include the Golgi apparatus, cell wall, actin cortex, axon growth cones etc.

Reply:
>If you want to avoid "dips" you can specify a very flat ellipsis and - essentially - perform a rolling_disk.

This is not quite the same as what I had in mind. By using the time dimension, it is possible to distinguish moving objects from stationary even if they have the same size.

Reply:
>Sidenote: If you have the scenario that the background illumination doesn't change over time, wouldn't it be much simpler to take a picture of the background and then subtract that one from each frame of the video? It will for sure be faster...

I fully agree. Unfortunately, people often forget to take a background image (usually done by de-focusing) and it is often difficult to obtain one with the exact same conditions later. For moving objects, a median projection can sometimes be used as a background image but in some cases this does not give the desired results.

Sidenote2: If you have a background image and want to correct for vignetting, it is best to divide by the background image in order to correct not only for the uneven background but also for the lower signal at the edges.

Reply:
@thawn I just realized that your second sidenote is a pretty good argument for why it is better to return the background instead of the image with background removed. One might not always want to just subtract the background. Nice!

Also, since it is now possible to roll ND-ellipsoids, I will resolve this conversation to keep track of the open discussions.